File: engine/language-server/src/main/scala/org/enso/languageserver/boot/resource/AsyncResourcesInitialization.scala
package org.enso.languageserver.boot.resource

import scala.concurrent.{ExecutionContext, Future}

/** Initializes resources in parallel.
  *
  * @param resources the list of resources to initialize
  * @param ec the execution context
  */
class AsyncResourcesInitialization(
  resources: Iterable[InitializationComponent]
)(implicit ec: ExecutionContext)
    extends InitializationComponent {

  /** @inheritdoc */
  override def init(): Future[InitializationComponent.Initialized.type] =
    Future
      .traverse(resources)(_.init())
      .map { _ => InitializationComponent.Initialized }
}

object AsyncResourcesInitialization {

  /** Create [[AsyncResourcesInitialization]] component.
    *
    * @param resources the list of resources to initialize
    * @param ec the execution context
    * @return new async initialization component
    */
  def apply(resources: InitializationComponent*)(implicit
    ec: ExecutionContext
  ): AsyncResourcesInitialization =
    new AsyncResourcesInitialization(resources)
}

File: engine/language-server/src/main/java/org/enso/languageserver/boot/resource/AsyncResourcesInitialization.java
package org.enso.languageserver.boot.resource;

import java.util.Arrays;
import java.util.concurrent.CompletableFuture;

/** Component that initializes resources in parallel. */
public class AsyncResourcesInitialization implements InitializationComponent {

  private final InitializationComponent[] resources;

  /**
   * Create async initialization component.
   *
   * @param resources the list of resources to initialize
   */
  public AsyncResourcesInitialization(InitializationComponent... resources) {
    this.resources = resources;
  }

  @Override
  public boolean isInitialized() {
    return Arrays.stream(resources).allMatch(InitializationComponent::isInitialized);
  }

  @Override
  public CompletableFuture<Void> init() {
    return CompletableFuture.allOf(
            Arrays.stream(resources)
                .map(
                    component ->
                        component.isInitialized()
                            ? CompletableFuture.completedFuture(null)
                            : component.init())
                .toArray(CompletableFuture<?>[]::new))
        .thenRun(() -> {});
  }
}
